File: Examples/ProductRecommendations/Obj-C/RecommendationsExample/ViewController.m
//
//  ViewController.m
//  RecommendationsExample
//
//  Copyright © 2016 Bazaarvoice. All rights reserved.
//

#import "ViewController.h"
#import "DemoCell.h"

@interface ViewController ()

@property (weak, nonatomic) IBOutlet BVProductRecommendationsCollectionView *recommendationsView;
@property NSArray*products;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.products = [NSArray array];
    
    [self.recommendationsView registerNib:[UINib nibWithNibName:@"DemoCell" bundle:nil]
               forCellWithReuseIdentifier:@"DemoCellIdentifier"];
    
    self.recommendationsView.delegate = self;
    self.recommendationsView.dataSource = self;
    
    [self loadRecommendations];
    
}

// Fetch the product recommenations from the API method on the BVRecommendationsCollectionView container.
- (void)loadRecommendations {
    
    BVRecommendationsRequest *request = [[BVRecommendationsRequest alloc] initWithLimit:20];
    
    [self.recommendationsView loadRequest:request completionHandler:^(NSArray * proudcts) {
        self.products = proudcts;
        [self.recommendationsView reloadData];
    } errorHandler:^(NSError * error) {
        NSLog(@"ERROR: %@", error.localizedDescription);
    }];
    
}

// Lays out the recommendationView with horitontal scrolling
- (void)viewWillLayoutSubviews{
    [super viewWillLayoutSubviews];
    
    UICollectionViewFlowLayout *layout = (UICollectionViewFlowLayout *)self.recommendationsView.collectionViewLayout;
    
    layout.scrollDirection = UICollectionViewScrollDirectionHorizontal;
    layout.itemSize = CGSizeMake(self.recommendationsView.bounds.size.width / 2, self.recommendationsView.bounds.size.width / 2);
    
    layout.minimumInteritemSpacing = 0;
    layout.minimumLineSpacing = 0;
}


#pragma mark  UICollectionViewDatasource

- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section{
    return self.products.count;
}

- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath{
    
    DemoCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier:@"DemoCellIdentifier" forIndexPath:indexPath];
    
    cell.bvProduct = [self.products objectAtIndex:indexPath.row];
    
    return cell;
    
}

#pragma mark UICollectionViewDelegate

- (void)collectionView:(UICollectionView *)collectionView didSelectItemAtIndexPath:(NSIndexPath *)indexPath{
    
    NSLog(@"Selected: %@", [self.products objectAtIndex:indexPath.row]);
    
}

@end

File: Examples/ProductRecommendations/Swift/RecommendationsExample/ViewController.swift
//
//  ViewController.swift
//  RecommendationsExample
//
//  Copyright © 2016 Bazaarvoice. All rights reserved.
//

import BVSDK

class ViewController: UIViewController, UICollectionViewDelegate, UICollectionViewDataSource {
    
    @IBOutlet weak var recommendationsView : BVProductRecommendationsCollectionView!
    var recommendations : [BVRecommendedProduct] = []
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        recommendationsView.register(UINib(nibName: "DemoCell", bundle: nil), forCellWithReuseIdentifier: "DemoCellIdentifier")
        
        self.recommendationsView.delegate = self
        self.recommendationsView.dataSource = self
        
        let request = BVRecommendationsRequest(limit: 20)
        
        self.recommendationsView.load(request, completionHandler: { (recommendations:[BVRecommendedProduct]) in
            
            self.recommendations = recommendations
            self.recommendationsView.reloadData()
            
        }) { (error) in
            // handle error case
            print("Error Loading Recommendations")
        }
        
    }

    override func viewWillLayoutSubviews() {
        super.viewWillLayoutSubviews()
        
        let layout = (recommendationsView.collectionViewLayout) as! UICollectionViewFlowLayout
        
        layout.scrollDirection = UICollectionViewScrollDirection.horizontal
        
        layout.itemSize = CGSize(width: recommendationsView.bounds.width / 2, height: recommendationsView.bounds.width / 2)
        
        layout.minimumInteritemSpacing = 0
        layout.minimumLineSpacing = 0
    }
    
    // MARK: UICollectionViewDataSource
    
    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return self.recommendations.count 
    }
    
    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
            
            let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "DemoCellIdentifier", for: indexPath as IndexPath) as! DemoCell
            
            let product = recommendations[indexPath.row]
            
            // Must set the bvProduct on each cell
            cell.bvRecommendedProduct = product
            
            return cell;
    }
    
    
    // MARK UICollectionViewDataSource
    
    func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
        print("Selected product: %@", recommendations[indexPath.row])
    }
    

}
